### PR TITLE
Remove a lot of synchronous code and simplify another a lot

### PR DESCRIPTION
**Pull request by luisholanda**

### CODE REVIEW

File: google-interface/src/cloudstorage/lib.js [hunk 1]
```diff
@@ -81,21 +80,26 @@ async function uploadFile(localFilename, destinationPath) {
...
-    gzip: false
-  };
+      const options = {
+        // The path to which the file should be downloaded, e.g. "./file.txt"
+        destination: destFilename,
+        gzip: false
+      };
+
+      // Downloads the file
+      bucket.file(srcFilename).download(options).then(reject)
```
Comment (megatron0000) [inline google-interface/src/cloudstorage/lib.js]:
```suggestion
      bucket.file(srcFilename).download(options).then(resolve)
```

File: google-interface/src/credentials/auth.js [hunk 2]
```diff
@@ -29,49 +29,39 @@ exports.getOAuth2ClientFromLocalCredentials = function getOAuth2ClientFromLocalC
...
-    fs.readFile(oauthClientCredentialsPath, (err, credentials) => {
-      if (err) {
-        err.message = 'Error loading client oauth credentials file: ' + err.message
-        return reject(err)
-      }
-      // Authorize a client with credentials
-      authorize(JSON.parse(credentials), oauthUserTokenPath, resolve, reject)
-    })
-  }
+  const  credentials = await fs.readFile(oauthClientCredentialsPath)
```
Comment (megatron0000) [inline google-interface/src/credentials/auth.js]: `await` ? So this function must be made `async`

File: google-interface/src/credentials/auth.js [hunk 3]
```diff
@@ -29,49 +29,39 @@ exports.getOAuth2ClientFromLocalCredentials = function getOAuth2ClientFromLocalC
...
+  return authorize(JSON.parse(credentials), oauthUserTokenPath)
 }
 
 /**
  * Create an OAuth2 client with the given credentials, and then execute the
  * given callback function.
  * @param {Object} credentials The authorization client credentials.
  */
-function authorize(credentials, userTokenPath, resolve, reject) {
+function authorize(credentials, userTokenPath) {
```
Comment (megatron0000) [inline google-interface/src/credentials/auth.js]:
```suggestion
async function authorize(credentials, userTokenPath) {
```

File: google-interface/src/credentials/auth.js [hunk 4]
```diff
@@ -84,22 +74,21 @@ function getNewToken(oAuth2Client, userTokenPath, resolve, reject) {
     output: process.stdout,
   })
 
-  rl.question('Enter the code from that page here: ', (code) => {
+  return rl.question('Enter the code from that page here: ', (code) => {
```
Comment (megatron0000) [inline google-interface/src/credentials/auth.js]: `rl.question()` does not return a Promise, but we still need to return one (with the OAuth2Client object as the "context"). So we need to revert to using `return new Promise((resolve, reject) => ...)`

File: google-interface/src/cloudstorage/index.js [hunk 5]
```diff
@@ -1,106 +1,109 @@
...
- * 
+ *
  * @param {*} courseId google-id
  * @param {*} courseWorkId google-id
  * @param {*} submissionId google-id
  * @param {*} localDirectory  path of the directory containing the student's submission files
  */
 async function uploadCourseWorkSubmissionFiles(courseId, courseWorkId, submissionId, localDirectory) {
-  const cloudDirectory = path.posix.join(
+  const cloudDirectory = path.join(
```
[the remark this answers is not in the document]
Comment (luisholanda) [reply google-interface/src/cloudstorage/index.js]: Humm, I think that we always run the code inside a *nix container, technically `path.posix` or `path` doesn't makes difference. I will keep the original `path.posix` to facilitate testing for the DOS guys.

File: google-interface/src/pubsub/index.js [hunk 6]
```diff
@@ -26,21 +26,21 @@ const listeners = []
...
 
 // Create an event handler to handle messages
-const messageHandler = message => {
+const messageHandler = async message => {
   /* console.log(`Received message ${message.id} (${new Date()}):`);
   console.log(`\tData: ${message.data}`);
   console.log(`\tAttributes: ${JSON.stringify(message.attributes)}\n`); */
 
-  listeners.forEach(listener => listener(message))
+  await Promise.all(listeners.forEach(listener => listener(message)))
```
[the remark this answers is not in the document]
Comment (luisholanda) [reply google-interface/src/pubsub/index.js]: Done.

File: google-interface/src/cloudstorage/index.js [hunk 7]
```diff
@@ -1,106 +1,109 @@
...
-  fs.readdirSync(dir).forEach(file => {
-    const filePath = path.join(dir, file)
-
-    if (fs.statSync(filePath).isDirectory()) {
-      allFilesSync(filePath).forEach(children => {
-        fileList.push(path.join(file, children))
-      })
-    } else {
-      fileList.push(file)
+const allFiles = async (dir, fileList = []) => {
```
[the remark this answers is not in the document]
Comment (luisholanda) [reply google-interface/src/cloudstorage/index.js]: Done.

File: backend/src/index.js [hunk 8]
```diff
@@ -4,27 +4,33 @@ const { submissionIsTurnedIn, assignGradeToSubmission, getSubmissionDriveFileIds
...
  * @param {*} searchName Filename to be searched (like "main.cpp")
  * @returns {string[]} Absolute paths of all matching files found
  */
-function findFileRecursive(rootDir, searchName) {
-  const filenames = readdirSync(rootDir)
-  const dirs = filenames.filter(filename => statSync(join(rootDir, filename)).isDirectory())
-  const files = filenames.filter(filename => statSync(join(rootDir, filename)).isFile())
+async function findFileRecursive(rootDir, searchName) {
+  const filenames = await readdir(rootDir)
+  const fileStats = await Promise.all(filename.map(file => stat(join(rootDir, file))))
```
[the remark this answers is not in the document]
Comment (luisholanda) [reply backend/src/index.js]: Done.

File: google-interface/src/cloudstorage/lib.js (hunk 1 above)
Comment (luisholanda) [reply google-interface/src/cloudstorage/lib.js]: Done.

File: google-interface/src/credentials/auth.js (hunk 2 above)
Comment (luisholanda) [reply google-interface/src/credentials/auth.js]: Done.

File: google-interface/src/credentials/auth.js (hunk 4 above)
Comment (luisholanda) [reply google-interface/src/credentials/auth.js]: Humm, so much time that I don't use callbacks, for me the function returned the result from the callback, from the [source](https://github.com/nodejs/node/blob/master/lib/readline.js#L293) of `question`, this is not the case. Fixing.

**Comment by luisholanda** [comment]
Fixed the issues. I've to learn to not write production code at 3am.

**Pull request merged**